### PR TITLE
Add Android test coverage

### PR DESCRIPTION
**Pull request by jonnyandrew**
Rebase of matrix-org/matrix-rich-text-editor#668

## Changes
- Publish Android test coverage to Codecov
- Refactor Android packaging
  - Package UI related code together
  - Package some internal code in the `internal` package

## [Codecov flags](https://app.codecov.io/github/matrix-org/matrix-rich-text-editor/pull/671/flags)
image

### CODE REVIEW

File: platforms/android/library/build.gradle [hunk 1]
```diff
@@ -4,6 +4,7 @@ plugins {
 
     // Rust
     id 'org.mozilla.rust-android-gradle.rust-android'
+    id("com.dicedmelon.gradle.jacoco-android") version "0.1.5"
```
Comment (jonnyandrew) [inline platforms/android/library/build.gradle:7]: Needed for Robolectric test coverage to be included in Jacoco report

File: .github/workflows/android.yml [hunk 2]
```diff
@@ -109,3 +112,17 @@ jobs:
           path : |
             ./**/build/reports/tests/**
             ./**/build/reports/androidTests/connected/**
+
+      - name: Upload unit test coverage to Codecov
+        uses: codecov/codecov-action@v3
+        with:
+          files: build/reports/jacoco/generateUnitTestCoverageReport/generateUnitTestCoverageReport.xml
+          fail_ci_if_error: false
+          flags: unittests-android, unittests
```
Comment (jonnyandrew) [inline .github/workflows/android.yml:121]: Note that we don't seem to need the Codecov token as this is a public repo using Github Actions. Can/should we remove the token for iOS/Rust? @langleyd

File: platforms/android/coverage.gradle [hunk 3]
```diff
@@ -1,37 +1,29 @@
-def excludes = [
-// dependency injection graph
-'**/*Module.*',
-'**/*Module*.*',
-
-// Framework entry points
-'**/*Activity*',
-'**/*Fragment*',
-'**/*Application*',
-'**/*AndroidService*',
```
Comment (jonnyandrew) [inline platforms/android/coverage.gradle:10]: This config had been copied across from Element Android but it's not needed in this code base

File: platforms/android/coverage.gradle [hunk 4]
```diff
@@ -77,16 +69,42 @@ task generateCoverageReport(type: JacocoReport) {
...
+    def projects = collectProjects { ['library'].contains(it.name) }
+    def executionDataPaths = [
+            "**/build/outputs/code_coverage/*AndroidTest/connected/**/coverage.ec",
+    ]
+    def excludes = commonExcludes + instrumentationTestExcludes
+    initializeReport(it, projects, excludes, executionDataPaths)
 }
 
 task unitTestsWithCoverage(type: GradleBuild) {
-    // the 7.1.3 android gradle plugin has a bug where enableTestCoverage generates invalid coverage
```
Comment (jonnyandrew) [inline platforms/android/coverage.gradle:84]: Fixed after updating to 7.2.2

File: .github/workflows/android.yml (hunk 2 above)
Comment (langleyd) [reply .github/workflows/android.yml:121]: I was getting some 404s sporadically that match [this issue](https://github.com/codecov/codecov-action/issues/557#issuecomment-1216749652), which form what I gathered is apparently due to codecov hitting their github rate limit in cases where the codecov token is not provided for public repos.

Anecdotally for me(as some others in that issue) adding the token resolved it.

Comment (jonnyandrew) [reply .github/workflows/android.yml:121]: [`ef87e69`](<https://github.com/matrix-org/matrix-rich-text-editor/commit/ef87e693f3663a12481b0981cc4149fe056d418c>)

**Review by langleyd — APPROVED** [review]
LGTM, thanks for that, quite a bit more involved than I anticipated!

**Pull request merged**